numba/tests/test_dyn_array.py

```diff
@@ -0,0 +1,362 @@
+        @nrtjit
+        def foo(n):
+            arr = np.empty(n)
+            for i in range(n):
+                arr[i] = i
+
+            return arr
+
+        n = 3
+        arr = foo(n)
+        np.testing.assert_equal(np.arange(n), arr)
+        self.assertEqual(arr.size, n)
+        self.assertEqual(arr.shape, (n,))
+        self.assertEqual(arr.dtype, np.dtype(np.float64))
+        self.assertEqual(arr.strides, (np.dtype(np.float64).itemsize,))
+        arr.fill(123)  # test writability
+        np.testing.assert_equal(123, arr)
+        del arr
+
+    def test_empty_2d(self):
+        def pyfunc(m, n):
+            arr = np.empty((m, n), np.int32)
+            for i in range(m):
+                for j in range(n):
+                    arr[i, j] = i + j
+
+            return arr
+
+        cfunc = nrtjit(pyfunc)
+        m = 4
+        n = 3
+        expected_arr = pyfunc(m, n)
+        got_arr = cfunc(m, n)
+        np.testing.assert_equal(expected_arr, got_arr)
+
+        self.assertEqual(expected_arr.size, got_arr.size)
+        self.assertEqual(expected_arr.shape, got_arr.shape)
+        self.assertEqual(expected_arr.strides, got_arr.strides)
+
+        del got_arr
+
+    def test_empty_3d(self):
+        def pyfunc(m, n, p):
+            arr = np.empty((m, n, p), np.int32)
+            for i in range(m):
+                for j in range(n):
+                    for k in range(p):
+                        arr[i, j, k] = i + j + k
+
+            return arr
+
+        cfunc = nrtjit(pyfunc)
+        m = 4
+        n = 3
+        p = 2
+        expected_arr = pyfunc(m, n, p)
+        got_arr = cfunc(m, n, p)
+        np.testing.assert_equal(expected_arr, got_arr)
+
+        self.assertEqual(expected_arr.size, got_arr.size)
+        self.assertEqual(expected_arr.shape, got_arr.shape)
+        self.assertEqual(expected_arr.strides, got_arr.strides)
+
+        del got_arr
+
+    def test_empty_2d_sliced(self):
+        def pyfunc(m, n, p):
+            arr = np.empty((m, n), np.int32)
+            for i in range(m):
+                for j in range(n):
+                    arr[i, j] = i + j
+
+            return arr[p]
+
+        cfunc = nrtjit(pyfunc)
+        m = 4
+        n = 3
+        p = 2
+        expected_arr = pyfunc(m, n, p)
+        got_arr = cfunc(m, n, p)
+        np.testing.assert_equal(expected_arr, got_arr)
+
+        self.assertEqual(expected_arr.size, got_arr.size)
+        self.assertEqual(expected_arr.shape, got_arr.shape)
+        self.assertEqual(expected_arr.strides, got_arr.strides)
+
+        del got_arr
+
+    def test_return_global_array(self):
+        y = np.ones(4, dtype=np.float32)
+        initrefct = sys.getrefcount(y)
+
+        def return_external_array():
+            return y
+
+        cfunc = nrtjit(return_external_array)
+        out = cfunc()
+
+        # out reference by cfunc
+        self.assertEqual(initrefct + 1, sys.getrefcount(y))
+
+        np.testing.assert_equal(y, out)
```

Comment:
Is `out` the same object as `y`? Is it backed by the same memory? If yes, then test mutating one effects the other as well?